pandas/core/groupby.py

```diff
@@ -1488,19 +1492,24 @@ def aggregate(self, values, how, axis=0):
                 raise NotImplementedError
             out_shape = (self.ngroups,) + values.shape[1:]
 
-        if is_numeric_dtype(values.dtype):
-            values = com.ensure_float(values)
+        upsample = self.ngroups > values.size
+
+        if _is_datetime_or_timedelta_dtype(values.dtype):
+            values = values.view('int64')
+
+        if (is_integer_dtype(values.dtype) and
+                how in _non_arithmetic_agg and not upsample):
```

Comment:
what ops don't pass this test? you have listed several that DO. can you explain.